pkg/controller/networkpolicy/clustergroup.go

```diff
@@ -333,25 +365,48 @@ func (n *NetworkPolicyController) GetAssociatedGroups(name, namespace string) ([
 	if !exists {
 		return nil, nil
 	}
-	groupObjs := make([]antreatypes.Group, 0, len(clusterGroups))
+	var groupObjs []antreatypes.Group
 	for _, g := range clusterGroups {
-		group, found, _ := n.internalGroupStore.Get(g)
-		if found {
-			groupObjs = append(groupObjs, *group.(*antreatypes.Group))
+		groupObjs = append(groupObjs, n.getAssociatedGroupsByName(g)...)
+	}
+	// Remove duplicates in the groupObj slice.
+	groupKeys, j := make(map[string]bool), 0
+	for _, g := range groupObjs {
+		if _, exists := groupKeys[g.Name]; !exists {
+			groupKeys[g.Name] = true
+			groupObjs[j] = g
+			j++
 		}
 	}
-	return groupObjs, nil
+	return groupObjs[:j], nil
+}
+
+// getAssociatedGroupsByName retrieves the internal Group and all it's parent Group objects
+// (if any) by Group name. This function assumes that there's no loop in Group reference.
+func (n *NetworkPolicyController) getAssociatedGroupsByName(grpName string) []antreatypes.Group {
+	var groups []antreatypes.Group
+	groupObj, found, _ := n.internalGroupStore.Get(grpName)
+	if found {
+		grp := groupObj.(*antreatypes.Group)
+		groups = append(groups, *grp)
+		parentGroupObjs, err := n.internalGroupStore.GetByIndex(store.ChildGroupIndex, grp.Name)
+		if err != nil {
+			klog.Errorf("Error retrieving parents of ClusterGroup %s: %v", grp.Name, err)
+		}
+		for _, p := range parentGroupObjs {
```

Comment:
Not necessarily. A child group can have multiple parents: 
cgA{selector: xx}, cgB{selector: yy},  cgC{childGroups: [cgA]}, cgD{childGroups: [cgA, cgB]}

Reply:
I meant the restriction "Currently only one level of 'nested-ness' is supported". Since p is a parent group, it cannot be another group's child, why we call `getAssociatedGroupsByName` for these parent groups?

Reply:
This is because I want to make getAssociateGroup and group membership functions extensible, so that they won't need to be modified when we increase the level of nested-ness: we would only need to modify the admission validations. For current usage, if you think there's too much overhead added to do an indexer call for each parent (which I feel is ok), then I can change it to be non-recursive.

Reply:
Not only the few overhead, I'm a little concerned about the assumption that there's no loop in Group reference. Today it's guarded by adminission validation that only one level of nested-ness is allowed, not sure how we ensure it later when the restriction is removed and whether we will have such validation according to recent discussion. I would prefer the code is extended to handle this case when it needs to and we are clear it's safe to be recursive.

Reply:
Okay. Done.